### PR TITLE
Dont run snapshot feeds in case cdc stopped for stale offset

### PR DESCRIPTION
**Pull request by rodireich**
## What
Stops all feeds in case global feed has exited with a `ConfigErrorException` exception.
This will cause a CDC connector to terminate immediately in case a saved CDC offset is no longer present on the server.

See [comment](https://github.com/airbytehq/airbyte-internal-issues/issues/10221#issuecomment-2410011398). This is the minimal solution that only solves the case we need.
## How
Adding logic inside `RootReader.read` which cancels the entire job if the CDC global partition had exited with Config error.
Running under `supervisorScope`, read block normally allows one feed to error out while continuing other feeds.

## User Impact
This change achieves one of the design goals of the WASS algorithm when reading CDC.

## Can this PR be safely reverted and rolled back?

- [ ] YES 💚
- [ ] NO ❌

### CODE REVIEW

File: airbyte-cdk/bulk/core/extract/src/main/kotlin/io/airbyte/cdk/read/RootReader.kt [hunk 1]
```diff
@@ -66,22 +69,32 @@ class RootReader(
...
             // Call listener hook.
             listener(feedJobs)
-            // Join on all stream feeds and collect caught exceptions.
-            val streamExceptions: Map<Stream, Throwable?> =
-                feeds.filterIsInstance<Stream>().associateWith {
-                    feedJobs[it]?.join()
-                    exceptions[it]
-                }
             // Join on all global feeds and collect caught exceptions.
             val globalExceptions: Map<Global, Throwable?> =
```
Comment (rodireich) [inline airbyte-cdk/bulk/core/extract/src/main/kotlin/io/airbyte/cdk/read/RootReader.kt:73]: Note: It makes sense to join `Global` feed(s) first as global runs before `Stream`.

File: airbyte-cdk/bulk/core/extract/src/main/kotlin/io/airbyte/cdk/read/RootReader.kt [hunk 2]
```diff
@@ -66,22 +69,32 @@ class RootReader(
...
                 feeds.filterIsInstance<Global>().associateWith {
                     feedJobs[it]?.join()
                     exceptions[it]
                 }
+
+            // Certain errors on the global feed cause a full stop to all stream reads
+            if (globalExceptions.values.filterIsInstance<ConfigErrorException>().isNotEmpty()) {
+                this@supervisorScope.cancel()
+            }
```
Comment (rodireich) [inline airbyte-cdk/bulk/core/extract/src/main/kotlin/io/airbyte/cdk/read/RootReader.kt:82]: Is there an error in global that shouldn't cause a full stop ?
I wasn't sure. Minimized cancellation to `ConfigErrorException` only

File: airbyte-cdk/bulk/core/extract/src/main/kotlin/io/airbyte/cdk/read/RootReader.kt [hunk 3]
```diff
@@ -66,22 +69,32 @@ class RootReader(
...
+
+            // Join on all stream feeds and collect caught exceptions.
+            val streamExceptions: Map<Stream, Throwable?> =
+                feeds.filterIsInstance<Stream>().associateWith {
+                    try {
+                        feedJobs[it]?.join()
+                        exceptions[it]
+                    } catch (_: CancellationException) {
+                        null
```
Comment (rodireich) [inline airbyte-cdk/bulk/core/extract/src/main/kotlin/io/airbyte/cdk/read/RootReader.kt:91]: cancelled jobs will throw exception when `joined.
There's no point in rethrowing these exceptions.

File: airbyte-cdk/bulk/core/extract/src/main/kotlin/io/airbyte/cdk/read/RootReader.kt (hunk 2 above)
Comment (akashkulk) [reply airbyte-cdk/bulk/core/extract/src/main/kotlin/io/airbyte/cdk/read/RootReader.kt:82]: Probably a good call. There might be transient errors for which we'd want to resume initial snapshots for

File: airbyte-cdk/bulk/core/extract/src/test/kotlin/io/airbyte/cdk/read/RootReaderIntegrationTest.kt [hunk 4]
```diff
@@ -194,6 +195,51 @@ class RootReaderIntegrationTest {
...
+            )
+        val testOutputConsumer = BufferingOutputConsumer(ClockFactory().fixed())
+        val rootReader =
+            RootReader(
+                stateManager,
+                slowHeartbeat,
+                excessiveTimeout,
+                testOutputConsumer,
+                listOf(
+                    ConfigErrorThrowingGlobalPartitionsCreatorFacrtory(
```
Comment (akashkulk) [inline airbyte-cdk/bulk/core/extract/src/test/kotlin/io/airbyte/cdk/read/RootReaderIntegrationTest.kt:214]: nit : ConfigErrorThrowingGlobalPartitionsCreatorFacrtory -> ConfigErrorThrowingGlobalPartitionsCreatorFactory

**Review by akashkulk — APPROVED** [review]

File: airbyte-cdk/bulk/core/extract/src/test/kotlin/io/airbyte/cdk/read/RootReaderIntegrationTest.kt (hunk 4 above)
Comment (rodireich) [reply airbyte-cdk/bulk/core/extract/src/test/kotlin/io/airbyte/cdk/read/RootReaderIntegrationTest.kt:214]: 🥵

**Pull request merged**